File: 算法竞赛进阶指南/GoDS (Go Data Structures)/src/0-segmenttree/library/RangeChminChmaxAddRangeSum/RangeChminChmaxAddRangeSum.py
#  https://judge.yosupo.jp/problem/range_chmin_chmax_add_range_sum
#  区间min/max/加法/区间和
#  0 left right x => 区间[left,right)的每个元素变为min(x,原值)
#  1 left right x => 区间[left,right)的每个元素变为max(x,原值)
#  2 left right x => 区间[left,right)的每个元素都加上x
#  3 left right => 求区间[left,right)的和

#  普通的线段树不能用了
#  需要用 SegmentTreeBeats


INF = int(1e18)


class SegmentTreeBeats:
    def __init__(self, n):
        self.n = n
        self.log = (n - 1).bit_length()
        self.size = 1 << self.log
        self.fmax = [-INF] * (2 * self.size)
        self.fmin = [INF] * (2 * self.size)
        self.smax = [-INF] * (2 * self.size)
        self.smin = [INF] * (2 * self.size)
        self.maxc = [0] * (2 * self.size)
        self.minc = [0] * (2 * self.size)
        self.sum = [0] * (2 * self.size)
        self.add = [0] * (2 * self.size)
        self.upd = [INF] * (2 * self.size)
        self.up = []
        self.down = []
        self.lt = [0] * (2 * self.size)
        self.rt = [0] * (2 * self.size)
        for i in range(self.size):
            self.lt[self.size + i] = i
            self.rt[self.size + i] = i + 1
        for i in range(self.size)[::-1]:
            self.lt[i] = self.lt[i << 1]
            self.rt[i] = self.rt[(i << 1) + 1]

    def build(self, arr):
        for i, a in enumerate(arr):
            self.fmax[self.size + i] = a
            self.fmin[self.size + i] = a
            self.maxc[self.size + i] = 1
            self.minc[self.size + i] = 1
            self.sum[self.size + i] = a
        for i in range(1, self.size)[::-1]:
            self._merge(i)

    def range_chmax(self, a, b, x):
        self.down.append(1)
        while self.down:
            k = self.down.pop()
            if b <= self.lt[k] or self.rt[k] <= a or x <= self.fmin[k]:
                continue
            if a <= self.lt[k] and self.rt[k] <= b and x < self.smin[k]:
                self.chmin_(k, x)
                continue
            self.down_propagate(k)
            self.up.append(k)
        self.up_merge()

    def range_chmin(self, a, b, x):
        self.down.append(1)
        while self.down:
            k = self.down.pop()
            if b <= self.lt[k] or self.rt[k] <= a or self.fmax[k] <= x:
                continue
            if a <= self.lt[k] and self.rt[k] <= b and self.smax[k] < x:
                self.chmax_(k, x)
                continue
            self.down_propagate(k)
            self.up.append(k)
        self.up_merge()

    def range_add(self, a, b, x):
        self.down.append(1)
        while self.down:
            k = self.down.pop()
            if b <= self.lt[k] or self.rt[k] <= a:
                continue
            if a <= self.lt[k] and self.rt[k] <= b:
                self.add_(k, x)
                continue
            self.down_propagate(k)
            self.up.append(k)
        self.up_merge()

    def range_update(self, a, b, x):
        self.down.append(1)
        while self.down:
            k = self.down.pop()
            if b <= self.lt[k] or self.rt[k] <= a:
                continue
            if a <= self.lt[k] and self.rt[k] <= b:
                self.update_(k, x)
                continue
            self.down_propagate(k)
            self.up.append(k)
        self.up_merge()

    def get_max(self, a, b):
        self.down.append(1)
        v = -INF
        while self.down:
            k = self.down.pop()
            if b <= self.lt[k] or self.rt[k] <= a:
                continue
            if a <= self.lt[k] and self.rt[k] <= b:
                v = max(v, self.fmax[k])
                continue
            self.down_propagate(k)
        return v

    def get_min(self, a, b):
        self.down.append(1)
        v = INF
        while self.down:
            k = self.down.pop()
            if b <= self.lt[k] or self.rt[k] <= a:
                continue
            if a <= self.lt[k] and self.rt[k] <= b:
                v = min(v, self.fmin[k])
                continue
            self.down_propagate(k)
        return v

    def get_sum(self, a, b):
        self.down.append(1)
        v = 0
        while self.down:
            k = self.down.pop()
            if b <= self.lt[k] or self.rt[k] <= a:
                continue
            if a <= self.lt[k] and self.rt[k] <= b:
                v += self.sum[k]
                continue
            self.down_propagate(k)
        return v

    def _merge(self, k):
        self.sum[k] = self.sum[2 * k] + self.sum[2 * k + 1]
        if self.fmax[2 * k] < self.fmax[2 * k + 1]:
            self.fmax[k] = self.fmax[2 * k + 1]
            self.maxc[k] = self.maxc[2 * k + 1]
            self.smax[k] = max(self.fmax[2 * k], self.smax[2 * k + 1])
        elif self.fmax[2 * k] > self.fmax[2 * k + 1]:
            self.fmax[k] = self.fmax[2 * k]
            self.maxc[k] = self.maxc[2 * k]
            self.smax[k] = max(self.smax[2 * k], self.fmax[2 * k + 1])
        else:
            self.fmax[k] = self.fmax[2 * k]
            self.maxc[k] = self.maxc[2 * k] + self.maxc[2 * k + 1]
            self.smax[k] = max(self.smax[2 * k], self.smax[2 * k + 1])
        if self.fmin[2 * k] > self.fmin[2 * k + 1]:
            self.fmin[k] = self.fmin[2 * k + 1]
            self.minc[k] = self.minc[2 * k + 1]
            self.smin[k] = min(self.fmin[2 * k], self.smin[2 * k + 1])
        elif self.fmin[2 * k] < self.fmin[2 * k + 1]:
            self.fmin[k] = self.fmin[2 * k]
            self.minc[k] = self.minc[2 * k]
            self.smin[k] = min(self.smin[2 * k], self.fmin[2 * k + 1])
        else:
            self.fmin[k] = self.fmin[2 * k]
            self.minc[k] = self.minc[2 * k] + self.minc[2 * k + 1]
            self.smin[k] = min(self.smin[2 * k], self.smin[2 * k + 1])

    def propagate(self, k):
        if self.size <= k:
            return  # ?
        if self.upd[k] != INF:
            self.update_(2 * k, self.upd[k])
            self.update_(2 * k + 1, self.upd[k])
            self.upd[k] = INF
            return
        if self.add[k]:
            self.add_(2 * k, self.add[k])
            self.add_(2 * k + 1, self.add[k])
            self.add[k] = 0
        if self.fmax[k] < self.fmax[2 * k]:
            self.chmax_(2 * k, self.fmax[k])
        if self.fmin[2 * k] < self.fmin[k]:
            self.chmin_(2 * k, self.fmin[k])
        if self.fmax[k] < self.fmax[2 * k + 1]:
            self.chmax_(2 * k + 1, self.fmax[k])
        if self.fmin[2 * k + 1] < self.fmin[k]:
            self.chmin_(2 * k + 1, self.fmin[k])

    def up_merge(self):
        while self.up:
            self._merge(self.up.pop())

    def down_propagate(self, k):
        self.propagate(k)
        self.down.append(2 * k)
        self.down.append(2 * k + 1)

    def update_(self, k, x):
        self.fmax[k] = x
        self.smax[k] = -INF
        self.fmin[k] = x
        self.smin[k] = INF
        self.maxc[k] = self.rt[k] - self.lt[k]
        self.minc[k] = self.rt[k] - self.lt[k]
        self.sum[k] = x * (self.rt[k] - self.lt[k])
        self.add[k] = 0
        self.upd[k] = x

    def add_(self, k, x):
        self.fmax[k] += x
        if self.smax[k] != -INF:
            self.smax[k] += x
        self.fmin[k] += x
        if self.smin[k] != INF:
            self.smin[k] += x
        self.sum[k] += x * (self.rt[k] - self.lt[k])
        if self.upd[k] != INF:
            self.upd[k] += x
        else:
            self.add[k] += x

    def chmax_(self, k, x):
        self.sum[k] += (x - self.fmax[k]) * self.maxc[k]
        if self.fmax[k] == self.fmin[k]:
            self.fmax[k] = x
            self.fmin[k] = x
        elif self.fmax[k] == self.smin[k]:
            self.fmax[k] = x
            self.smin[k] = x
        else:
            self.fmax[k] = x
        if self.upd[k] != INF and x < self.upd[k]:
            self.upd[k] = x

    def chmin_(self, k, x):
        self.sum[k] += (x - self.fmin[k]) * self.minc[k]
        if self.fmin[k] == self.fmax[k]:
            self.fmin[k] = x
            self.fmax[k] = x
        elif self.fmin[k] == self.smax[k]:
            self.fmin[k] = x
            self.smax[k] = x
        else:
            self.fmin[k] = x
        if self.upd[k] != INF and self.upd[k] < x:
            self.upd[k] = x


import sys

input = sys.stdin.buffer.readline

n, q = map(int, input().split())
nums = tuple(map(int, input().split()))

stb = SegmentTreeBeats(n)
stb.build(nums)

res = []
for _ in range(q):
    op, *args = map(int, input().split())
    if op == 0:
        left, right, x = args
        stb.range_chmin(left, right, x)
    elif op == 1:
        left, right, x = args
        stb.range_chmax(left, right, x)
    elif op == 2:
        left, right, x = args
        stb.range_add(left, right, x)
    else:
        left, right = args
        res.append(stb.get_sum(left, right))

print("\n".join(map(str, res)))

File: 算法竞赛进阶指南/GoDS (Go Data Structures)/src/0-segmenttree/library/RangeChminChmaxAddRangeSum/RangeChminChmaxAddRangeSum.go
// https://judge.yosupo.jp/problem/range_chmin_chmax_add_range_sum
// 区间min/max/加法/区间和
// 0 left right x => 区间[left,right)的每个元素变为min(x,原值)
// 1 left right x => 区间[left,right)的每个元素变为max(x,原值)
// 2 left right x => 区间[left,right)的每个元素都加上x
// 3 left right => 求区间[left,right)的和
// !0<=left<right<=n
// 普通的线段树不能用了
// 需要用 SegmentTreeBeats

package main

import (
	"bufio"
	"fmt"
	"math/bits"
	"os"
)

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n, q int
	fmt.Fscan(in, &n, &q)

	nums := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Fscan(in, &nums[i])
	}
	tree := NewSegmentTreeBeats(n)
	tree.Build(nums)

	for i := 0; i < q; i++ {
		var op, left, right, x int
		fmt.Fscan(in, &op)
		if op == 0 {
			fmt.Fscan(in, &left, &right, &x)
			tree.RangeChmin(left, right, x)
		} else if op == 1 {
			fmt.Fscan(in, &left, &right, &x)
			tree.RangeChmax(left, right, x)
		} else if op == 2 {
			fmt.Fscan(in, &left, &right, &x)
			tree.RangeAdd(left, right, x)
		} else {
			fmt.Fscan(in, &left, &right)
			fmt.Fprintln(out, tree.GetSum(left, right))
		}
	}

}

const INF int = 1e18

type SegmentTreeBeats struct {
	n, log, size                                      int
	fmax, fmin, smax, smin, maxc, minc, add, upd, sum []int
	up, down, lt, rt                                  []int
}

func NewSegmentTreeBeats(n int) *SegmentTreeBeats {
	log := bits.Len(uint(n - 1))
	size := 1 << uint(log)
	fmax := make([]int, 2*size)
	fmin := make([]int, 2*size)
	smax := make([]int, 2*size)
	smin := make([]int, 2*size)
	maxc := make([]int, 2*size)
	minc := make([]int, 2*size)
	add := make([]int, 2*size)
	upd := make([]int, 2*size)
	sum := make([]int, 2*size)
	for i := range fmax {
		fmax[i] = -INF
		fmin[i] = INF
		smax[i] = -INF
		smin[i] = INF
		upd[i] = INF
	}
	up := make([]int, 0, 2*size)
	down := make([]int, 0, 2*size)
	lt := make([]int, 2*size)
	rt := make([]int, 2*size)
	for i := 0; i < size; i++ {
		lt[i+size] = i
		rt[i+size] = i + 1
	}
	for i := size - 1; i > -1; i-- {
		lt[i] = lt[i<<1]
		rt[i] = rt[i<<1|1]
	}

	return &SegmentTreeBeats{
		n:    n,
		log:  log,
		size: size,
		fmax: fmax,
		fmin: fmin,
		smax: smax,
		smin: smin,
		maxc: maxc,
		minc: minc,
		add:  add,
		upd:  upd,
		sum:  sum,
		up:   up,
		down: down,
		lt:   lt,
		rt:   rt,
	}
}

func (stb *SegmentTreeBeats) Build(arr []int) {
	for i, a := range arr {
		stb.fmax[i+stb.size] = a
		stb.fmin[i+stb.size] = a
		stb.sum[i+stb.size] = a
		stb.maxc[i+stb.size] = 1
		stb.minc[i+stb.size] = 1
	}

	for i := stb.size - 1; i > 0; i-- {
		stb.merge(i)
	}
}

// RangeAdd updates all elements in [left, right) by x.
//  0 <= left < right <= n
func (stb *SegmentTreeBeats) RangeChmax(left, right, x int) {
	stb.down = append(stb.down, 1)
	for len(stb.down) > 0 {
		k := stb.down[len(stb.down)-1]
		stb.down = stb.down[:len(stb.down)-1]
		if right <= stb.lt[k] || stb.rt[k] <= left || x <= stb.fmin[k] {
			continue
		}
		if left <= stb.lt[k] && stb.rt[k] <= right && x < stb.smin[k] {
			stb.chmin(k, x)
			continue
		}
		stb.downPropagate(k)
		stb.up = append(stb.up, k)
	}
	stb.upMerge()
}

// RangeChmin updates all elements in [left, right) to min(x, a[i])
//  0 <= left < right <= n
func (stb *SegmentTreeBeats) RangeChmin(left, right, x int) {
	stb.down = append(stb.down, 1)
	for len(stb.down) > 0 {
		k := stb.down[len(stb.down)-1]
		stb.down = stb.down[:len(stb.down)-1]
		if right <= stb.lt[k] || stb.rt[k] <= left || stb.fmax[k] <= x {
			continue
		}
		if left <= stb.lt[k] && stb.rt[k] <= right && stb.smax[k] < x {
			stb.chmax(k, x)
			continue
		}
		stb.downPropagate(k)
		stb.up = append(stb.up, k)
	}
	stb.upMerge()
}

// RangeAdd adds x to all elements in [left, right)
//  0 <= left < right <= n
func (stb *SegmentTreeBeats) RangeAdd(left, right, x int) {
	stb.down = append(stb.down, 1)
	for len(stb.down) > 0 {
		k := stb.down[len(stb.down)-1]
		stb.down = stb.down[:len(stb.down)-1]
		if right <= stb.lt[k] || stb.rt[k] <= left {
			continue
		}
		if left <= stb.lt[k] && stb.rt[k] <= right {
			stb.add_(k, x)
			continue
		}
		stb.downPropagate(k)
		stb.up = append(stb.up, k)
	}
	stb.upMerge()
}

// RangeUpdate updates all elements in [left, right) to x.
//  0 <= left < right <= n
func (stb *SegmentTreeBeats) RangeUpdate(left, right, x int) {
	stb.down = append(stb.down, 1)
	for len(stb.down) > 0 {
		k := stb.down[len(stb.down)-1]
		stb.down = stb.down[:len(stb.down)-1]
		if right <= stb.lt[k] || stb.rt[k] <= left {
			continue
		}
		if left <= stb.lt[k] && stb.rt[k] <= right {
			stb.update(k, x)
			continue
		}
		stb.downPropagate(k)
		stb.up = append(stb.up, k)
	}
	stb.upMerge()
}

//  0 <= left < right <= n
func (stb *SegmentTreeBeats) GetMax(left, right int) int {
	stb.down = append(stb.down, 1)
	v := -INF
	for len(stb.down) > 0 {
		k := stb.down[len(stb.down)-1]
		stb.down = stb.down[:len(stb.down)-1]
		if right <= stb.lt[k] || stb.rt[k] <= left {
			continue
		}
		if left <= stb.lt[k] && stb.rt[k] <= right {
			v = max(v, stb.fmax[k])
			continue
		}
		stb.downPropagate(k)
	}
	return v
}

//  0 <= left < right <= n
func (stb *SegmentTreeBeats) GetMin(left, right int) int {
	stb.down = append(stb.down, 1)
	v := INF
	for len(stb.down) > 0 {
		k := stb.down[len(stb.down)-1]
		stb.down = stb.down[:len(stb.down)-1]
		if right <= stb.lt[k] || stb.rt[k] <= left {
			continue
		}
		if left <= stb.lt[k] && stb.rt[k] <= right {
			v = min(v, stb.fmin[k])
			continue
		}
		stb.downPropagate(k)
	}
	return v
}

//  0 <= left < right <= n
func (stb *SegmentTreeBeats) GetSum(left, right int) int {
	stb.down = append(stb.down, 1)
	var v int
	for len(stb.down) > 0 {
		k := stb.down[len(stb.down)-1]
		stb.down = stb.down[:len(stb.down)-1]
		if right <= stb.lt[k] || stb.rt[k] <= left {
			continue
		}
		if left <= stb.lt[k] && stb.rt[k] <= right {
			v += stb.sum[k]
			continue
		}
		stb.downPropagate(k)
	}
	return v
}

func (stb *SegmentTreeBeats) merge(k int) {
	stb.sum[k] = stb.sum[k<<1] + stb.sum[k<<1|1]

	if stb.fmax[k<<1] < stb.fmax[k<<1|1] {
		stb.fmax[k] = stb.fmax[k<<1|1]
		stb.maxc[k] = stb.maxc[k<<1|1]
		stb.smax[k] = max(stb.fmax[k<<1], stb.smax[k<<1|1])
	} else if stb.fmax[k<<1] > stb.fmax[k<<1|1] {
		stb.fmax[k] = stb.fmax[k<<1]
		stb.maxc[k] = stb.maxc[k<<1]
		stb.smax[k] = max(stb.smax[k<<1], stb.fmax[k<<1|1])
	} else {
		stb.fmax[k] = stb.fmax[k<<1]
		stb.maxc[k] = stb.maxc[k<<1] + stb.maxc[k<<1|1]
		stb.smax[k] = max(stb.smax[k<<1], stb.smax[k<<1|1])
	}

	if stb.fmin[k<<1] > stb.fmin[k<<1|1] {
		stb.fmin[k] = stb.fmin[k<<1|1]
		stb.minc[k] = stb.minc[k<<1|1]
		stb.smin[k] = min(stb.fmin[k<<1], stb.smin[k<<1|1])
	} else if stb.fmin[k<<1] < stb.fmin[k<<1|1] {
		stb.fmin[k] = stb.fmin[k<<1]
		stb.minc[k] = stb.minc[k<<1]
		stb.smin[k] = min(stb.smin[k<<1], stb.fmin[k<<1|1])
	} else {
		stb.fmin[k] = stb.fmin[k<<1]
		stb.minc[k] = stb.minc[k<<1] + stb.minc[k<<1|1]
		stb.smin[k] = min(stb.smin[k<<1], stb.smin[k<<1|1])
	}
}

func (stb *SegmentTreeBeats) propagate(k int) {
	if k >= stb.size {
		return
	}

	if stb.upd[k] != INF {
		stb.update(k<<1, stb.upd[k])
		stb.update(k<<1|1, stb.upd[k])
		stb.upd[k] = INF
		return
	}

	if stb.add[k] != 0 {
		stb.add_(k<<1, stb.add[k])
		stb.add_(k<<1|1, stb.add[k])
		stb.add[k] = 0
	}

	if stb.fmax[k] < stb.fmax[k<<1] {
		stb.chmax(k<<1, stb.fmax[k])
	}
	if stb.fmin[k<<1] < stb.fmin[k] {
		stb.chmin(k<<1, stb.fmin[k])
	}
	if stb.fmax[k] < stb.fmax[k<<1|1] {
		stb.chmax(k<<1|1, stb.fmax[k])
	}
	if stb.fmin[k<<1|1] < stb.fmin[k] {
		stb.chmin(k<<1|1, stb.fmin[k])
	}
}

func (stb *SegmentTreeBeats) upMerge() {
	for len(stb.up) > 0 {
		popped := stb.up[len(stb.up)-1]
		stb.up = stb.up[:len(stb.up)-1]
		stb.merge(popped)
	}
}

func (stb *SegmentTreeBeats) downPropagate(k int) {
	stb.propagate(k)
	stb.down = append(stb.down, k<<1, k<<1|1)
}

func (stb *SegmentTreeBeats) update(k, x int) {
	stb.fmax[k] = x
	stb.fmin[k] = x
	stb.smax[k] = -INF
	stb.smin[k] = INF
	stb.maxc[k] = stb.rt[k] - stb.lt[k]
	stb.minc[k] = stb.rt[k] - stb.lt[k]
	stb.upd[k] = x
	stb.add[k] = 0
	stb.sum[k] = x * (stb.rt[k] - stb.lt[k])
}

func (stb *SegmentTreeBeats) add_(k, x int) {
	stb.fmax[k] += x
	if stb.smax[k] != -INF {
		stb.smax[k] += x
	}
	stb.fmin[k] += x
	if stb.smin[k] != INF {
		stb.smin[k] += x
	}
	stb.sum[k] += x * (stb.rt[k] - stb.lt[k])
	if stb.upd[k] != INF {
		stb.upd[k] += x
	} else {
		stb.add[k] += x
	}
}

func (stb *SegmentTreeBeats) chmax(k, x int) {
	stb.sum[k] += (x - stb.fmax[k]) * stb.maxc[k]
	if stb.fmax[k] == stb.fmin[k] {
		stb.fmax[k] = x
		stb.fmin[k] = x
	} else if stb.fmax[k] == stb.smin[k] {
		stb.fmax[k] = x
		stb.smin[k] = x
	} else {
		stb.fmax[k] = x
	}
	if stb.upd[k] != INF && x < stb.upd[k] {
		stb.upd[k] = x
	}
}

func (stb *SegmentTreeBeats) chmin(k, x int) {
	stb.sum[k] += (x - stb.fmin[k]) * stb.minc[k]
	if stb.fmin[k] == stb.fmax[k] {
		stb.fmin[k] = x
		stb.fmax[k] = x
	} else if stb.fmin[k] == stb.smax[k] {
		stb.fmin[k] = x
		stb.smax[k] = x
	} else {
		stb.fmin[k] = x
	}
	if stb.upd[k] != INF && stb.upd[k] < x {
		stb.upd[k] = x
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
